### PR TITLE
client-go/util/jsonpath: resolve #16707 by outputting json for non-primitive types

### LINKED ISSUE #16707: jsonpath template output should be json
I would expect the output to be json. @kubernetes/kubectl 

$ kubectl get no -o jsonpath="{.items}"  
[map[kind:Node apiVersion:v1 metadata:map[uid:58f92fa4-81c3-11e5-8109-42010af0000b resourceVersion:3018 creationTimestamp:2015-11-03T00:39:31Z labels:map[kubernetes.io/hostname:k-5-master] name:k-5-master selfLink:/api/v1/nodes/k-5-master] spec:map[podCIDR:10.244.0.0/24 externalID:18000990123038728740 providerID:gce://mikedanese-k8s/us-central1-b/k-5-master unschedulable:true] status:map[nodeInfo:map[machineID: systemUUID:5A5B3748-A867-EBA7-EC86-23FEAA14660D bootID:6e05e906-0ca4-4ee3-927e-a36798084cb6 kernelVersion:3.16.0-0.bpo.4-amd64 osImage:Debian GNU/Linux 7 (wheezy) containerRuntimeVersion:docker://1.8.3 kubeletVersion:v1.2.0-alpha.2.626+8fc36e6068334a-dirty kubeProxyVersion:v1.2.0-alpha.2.626+8fc36e6068334a-dirty] capacity:map[cpu:2 memory:7679608Ki pods:40] conditions:[map[type:Ready status:True lastHeartbeatTime:2015-11-03T02:38:07Z lastTransitionTime:2015-11-03T00:39:31Z reason:KubeletReady message:kubelet is posting ready status] map[type:OutOfDisk status:False lastHeartbeatTime:2015-11-03T02:38:07Z lastTransitionTime:2015-11-03T00:39:31Z reason:KubeletHasSufficientDisk message:kubelet has sufficient disk space available]] addresses:[map[type:InternalIP address:10.240.0.11] map[type:ExternalIP address:104.197.106.102]] daemonEndpoints:map[kubeletEndpoint:map[Port:10250]]]] map[spec:map[podCIDR:10.244.2.0/24 externalID:14275614848584897735 providerID:gce://mikedanese-k8s/us-central1-b/k-5-minion-0jkl] status:map[daemonEndpoints:map[kubeletEndpoint:map[Port:10250]] nodeInfo:map[osImage:Debian GNU/Linux 7 (wheezy) containerRuntimeVersion:docker://1.8.3 kubeletVersion:v1.2.0-alpha.2.626+8fc36e6068334a-dirty kubeProxyVersion:v1.2.0-alpha.2.626+8fc36e6068334a-dirty machineID: systemUUID:0EAF030D-B18B-57C9-9616-4008F51DEAF4 bootID:84fbf71d-959d-46e5-8cd2-8a62859508c0 kernelVersion:3.16.0-0.bpo.4-amd64] capacity:map[cpu:2 memory:7679608Ki pods:40] conditions:[map[type:Ready status:True lastHeartbeatTime:2015-11-03T02:38:07Z lastTransitionTime:2015-11-03T00:40:52Z reason:KubeletReady message:kubelet is posting ready status] map[type:OutOfDisk status:False lastHeartbeatTime:2015-11-03T02:38:07Z lastTransitionTime:2015-11-03T00:40:15Z reason:KubeletHasSufficientDisk message:kubelet has sufficient disk space available]] addresses:[map[type:InternalIP address:10.240.0.32] map[address:130.211.149.36 type:ExternalIP]]] kind:Node apiVersion:v1 metadata:map[uid:73548e00-81c3-11e5-8109-42010af0000b resourceVersion:3019 creationTimestamp:2015-11-03T00:40:15Z labels:map[kubernetes.io/hostname:k-5-minion-0jkl] name:k-5-minion-0jkl selfLink:/api/v1/nodes/k-5-minion-0jkl]] map[kind:Node apiVersion:v1 metadata:map[name:k-5-minion-1mq6 selfLink:/api/v1/nodes/k-5-minion-1mq6 uid:6dfcefcc-81c3-11e5-8109-42010af0000b resourceVersion:3017 creationTimestamp:2015-11-03T00:40:06Z labels:map[kubernetes.io/hostname:k-5-minion-1mq6]] spec:map[podCIDR:10.244.1.0/24 externalID:2970383975702131324 providerID:gce://mikedanese-k8s/us-central1-b/k-5-minion-1mq6] status:map[conditions:[map[message:kubelet is posting ready status type:Ready status:True lastHeartbeatTime:2015-11-03T02:38:05Z lastTransitionTime:2015-11-03T00:40:42Z reason:KubeletReady] map[lastHeartbeatTime:2015-11-03T02:38:05Z lastTransitionTime:2015-11-03T00:40:06Z reason:KubeletHasSufficientDisk message:kubelet has sufficient disk space available type:OutOfDisk status:False]] addresses:[map[type:InternalIP address:10.240.0.31] map[type:ExternalIP address:108.59.86.162]] daemonEndpoints:map[kubeletEndpoint:map[Port:10250]] nodeInfo:map[kernelVersion:3.16.0-0.bpo.4-amd64 osImage:Debian GNU/Linux 7 (wheezy) containerRuntimeVersion:docker://1.8.3 kubeletVersion:v1.2.0-alpha.2.626+8fc36e6068334a-dirty kubeProxyVersion:v1.2.0-alpha.2.626+8fc36e6068334a-dirty machineID: systemUUID:B2C51EC4-7CB7-62DE-3005-40D3C3C4AA7E bootID:84635e0a-b6b8-4c85-b5aa-354b0f93a356] capacity:map[memory:7679608Ki pods:40 cpu:2]]] map[kind:Node apiVersion:v1 metadata:map[labels:map[kubernetes.io/hostname:k-5-minion-j8s2] name:k-5-minion-j8s2 selfLink:/api/v1/nodes/k-5-minion-j8s2 uid:74dff725-81c3-11e5-8109-42010af0000b resourceVersion:3016 creationTimestamp:2015-11-03T00:40:18Z] spec:map[providerID:gce://mikedanese-k8s/us-central1-b/k-5-minion-j8s2 podCIDR:10.244.3.0/24 externalID:15940688070743728911] status:map[daemonEndpoints:map[kubeletEndpoint:map[Port:10250]] nodeInfo:map[systemUUID:4FEBAB9C-F20C-1EAE-939D-7CF3E8A98E34 bootID:70059938-8e08-4202-a2c2-4b6658e0f88c kernelVersion:3.16.0-0.bpo.4-amd64 osImage:Debian GNU/Linux 7 (wheezy) containerRuntimeVersion:docker://1.8.3 kubeletVersion:v1.2.0-alpha.2.626+8fc36e6068334a-dirty kubeProxyVersion:v1.2.0-alpha.2.626+8fc36e6068334a-dirty machineID:] capacity:map[cpu:2 memory:7679608Ki pods:40] conditions:[map[message:kubelet is posting ready status type:Ready status:True lastHeartbeatTime:2015-11-03T02:38:02Z lastTransitionTime:2015-11-03T00:40:56Z reason:KubeletReady] map[lastHeartbeatTime:2015-11-03T02:38:02Z lastTransitionTime:2015-11-03T00:40:18Z reason:KubeletHasSufficientDisk message:kubelet has sufficient disk space available type:OutOfDisk status:False]] addresses:[map[type:InternalIP address:10.240.0.33] map[address:104.154.40.190 type:ExternalIP]]]]]

### LINKED ISSUE #67268: Human readable status from conditions for a CRD using additionalPrinterColumns
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:
When a CRD contains status conditions in a pattern similar to PodCondition there is not a good way to produce human readable output of the current status in kubectl. AdditionalPrinterColumns is useful for improving kubectl UX for data which can be pulled directly from the CRD but status conditions are generally stored as a list which requires additional processing to be human readable.

**What you expected to happen**:
I'd like the ability to select a single status condition's value for printing in a column via additionalPrinterColumns.

**How to reproduce it (as minimally and precisely as possible)**:
I have a CRD which defines the following schema under the property "status":
```go
type FooConditionType string

type FooCondition struct {
    Type FooConditionType
    // k8s.io/api/core/v1 ConditionStatus
    Status corev1.ConditionStatus
}

type FooStatus struct {
    Conditions []FooCondition
}

```
I am unable to provide a 'status' column for kubectl get based on a condition value e.g. I'd like to show ConditionReady's value in a READY column.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.11.1

### PR DESCRIPTION
**Pull request by pjferrell**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
When printing JSONPATH results, this changes the display format of non-primitive types from Go-syntax to JSON.  This improves readability for maps/slices and simplifies processing outputs in other tools.

**Which issue(s) this PR fixes**:

Fixes #16707
Fixes #67268

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
jsonpath support in kubectl / client-go serializes complex types (maps / slices / structs) as json instead of Go-syntax.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by brianpursley** [comment]
Hi @pjferrell 
Thanks for the PR.

if you edit your original message to change `Fixes # 16707` to `Fixes #16707` (remove the space) it will turn that into a link to the issue which will help the reviewers.

File: staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go
Comment (brianpursley) [inline staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go:128]: Do you want this change to apply for slice and struct?  I thought the issue was only about map.  I guess I would need to see some unit tests around those types to understand how the output would be different.

Comment (pjferrell) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go:128]: Based on the comments in #16707 (@thockin last comment and others) indicated it'd be favorable to address all non-primitive types which could be represented by json.  Go-style output of structs does not contain attribute/key names and slices are printed without comma separators, which makes passing these outputs to other tools more difficult.

Comment (pjferrell) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go:128]: The current test failure is an example jsonpath output where the slice was previously output is Go-style instead of json.

Expected was the old output, and actual is the proper json representation of the slice.
```
--- FAIL: TestTokenOutput (0.00s)
    --- PASS: TestTokenOutput/JSON_output (0.00s)
    --- PASS: TestTokenOutput/YAML_output (0.00s)
    --- PASS: TestTokenOutput/Go_template_output (0.00s)
    --- PASS: TestTokenOutput/text_output (0.00s)
    --- FAIL: TestTokenOutput/jsonpath_output (0.00s)
        token_test.go:449: failed TestTokenOutput:
            
            expected:
            abcdef.1234567890123456 [system:bootstrappers:kubeadm:default-node-token]
            
            actual:
            abcdef.1234567890123456 ["system:bootstrappers:kubeadm:default-node-token"]
```

Comment (brianpursley) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go:128]: Can you add a couple unit tests to cover the slice and struct scenarios.
Something like this might work to cover both (slice with multiple structs):

```
{range .items[1]}{.status.addresses} {end}
```

I'm just thinking it will be good to have this to show the expected output and prevent any future regression.

**Comment by brianpursley** [comment]
Suggestion on your release note, you might want to word it in a way that assumes it will be in a long list of changes.  When you say "this fix" it won't mean anything to someone reading the Kubernetes release notes because it is out of context.  

> When using the JSONPATH output option in kubectl (or client-go/util/jsonpath package), complex types (maps / slices / structs) were previously displayed as Go-syntax.  This fix changes the behavior to return json for non-primitive types.

I suggest rewording it to a single sentence if possible and just describe the new kubectl behavior without making any reference to this PR or the Issue.

File: staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go
Comment (pjferrell) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go]: I've added several test cases for each primitive type.  While adding test cases, I resolved some missing/broken filter behavior that was skipping certain types / unstable sorting.

**Comment by brianpursley** [comment]
Wow this is getting to be a pretty big PR compared to the original commit.  Maybe consider breaking the work up into multiple PRs will help the reviewers be able to approve more easily.

**Comment by pjferrell** [comment]
@brianpursley I pulled out the code for providing a stable/sorted jsonpath recursion/iteration (primarily affects recursion of maps).

The remaining code is the minimal changes to achieve:
1. client-go/util/jsonpath: output non-primitives (slice/struct/map) as json
2. kubectl: add --output=jsonpath-as-json to print a json list of results (ex. slice of Ingresses, slice of Containers, slice of .metadata.name, etc)
3. kubeadm: fix test case that was expecting Go-style output of slice

I'll create a new PR for stable/sorted jsonpath recursion/iteration once this PR is merged.  It will only affect client-go/util/jsonpath package

**Comment by pjferrell** [comment]
Related PRs that were closed:
https://github.com/kubernetes/kubernetes/pull/26099
https://github.com/kubernetes/kubernetes/pull/33890

**Comment by neolit123** [comment]
/approve
for cmd/kubeadm

/area kubectl
/priority backlog
(given this has been an issue for a while)

**Comment by lavalamp** [comment]
As long as this is a client-side change, which it seems to be, I'm fine with it (thanks for all the tests). I think someone from SIG CLI should sign off on it, then I'm happy to approve for the api machinery OWNERs files.

/assign @pwittrock @seans3 @apelisse

**Comment by pjferrell** [comment]
@lavalamp It does change the default output behavior for complex objects to json instead of go-syntax.  IMO it would be a positive change, but it could break a script expecting go-style output ("1 2 3" vs "[1,2,3]").  I fixed all kubernetes test cases (3) that used the go-style output.

The cli flag provides an output option to display a json slice of results (consistent with how other jsonpath implementations output).

**Comment by lavalamp** [comment]
Yeah, I think that makes sense, we just need SIG CLI to agree.

**Comment by apelisse** [comment]
I see a couple of comments from informed people that make sense to me:

From https://github.com/kubernetes/kubernetes/pull/33890#issuecomment-251278083:
> Single values are used to provide lists of arguments. Changing the format would break all scripts using it.
>
> The printing of complex values is a bug, and can be fixed.

From https://github.com/kubernetes/kubernetes/pull/33890#issuecomment-252347667:
> The go type dumps are borderline useless. We can either say JSON or we can
define our own format.

**Comment by lavalamp** [comment]
/approve
/lgtm

OK that's good enough for me

**Review by brianpursley — APPROVED** [review]

**Comment by thockin** [comment]
This is awesome!

Now how do I format a JSON list without a trailing comma?

```
kubectl get nodes -o jsonpath='{"{"}{range .items[*]}"{.metadata.name}": {.metadata.labels},{end}{"}"}' | jq .
parse error: Expected another key-value pair at line 1, column 1582
```

:P

**Comment by brianpursley** [comment]
Its probably not what you're looking for, but you *could* use sed to remove it...
```
kubectl get nodes -o jsonpath='{"{"}{range .items[*]}"{.metadata.name}": {.metadata.labels},{end}{"}"}' | sed 's/\(.*\),/\1/' | jq
```

Are you thinking there should be some kind of conditional rendering? ...maybe some new thing like `{separator ","}` that conditionally renders only when it is a separating range iterations? 

Is jsonpath an actual spec?

Also, if it were able to render \b, then you could add a `\b ` after the last `{end}` to "backspace away" the last comma.  A little hacky but might be a simple way to do it if it actually worked.

**Comment by lavalamp** [comment]
IMO just use jq if you want to do that, it's much simpler.

To do it without jq, I would do something exceedingly stupid, e.g.

```sh
$ echo "{x,y,}" | (head -c -3; echo "}")
{x,y}
```

**Comment by lavalamp** [comment]
I haven't tested it but a jq script should look something like this:

`jq '[.items[*] | {.metadata.name, .metadata.labels}]'`

(Yes, I feel guilty for being this nerdy about jq)

**Pull request merged**
